lib/Sema/TypeCheckType.cpp

```diff
@@ -862,7 +858,16 @@ Type TypeResolution::applyUnboundGenericArguments(
         module, loc, noteLoc,
         UnboundGenericType::get(decl, parentTy, getASTContext()),
         genericSig.getGenericParams(), genericSig.getRequirements(),
-        QueryTypeSubstitutionMap{contextualSubs});
+        [&](SubstitutableType *type) -> Type {
+          auto result = QueryTypeSubstitutionMap{subs}(type);
+          if (result->hasTypeParameter()) {
+            if (const auto contextSig = getGenericSignature()) {
+              auto *genericEnv = contextSig.getGenericEnvironment();
+              return genericEnv->mapTypeIntoContext(result);
+            }
+          }
+          return result;
+        });
```

Comment:
Huh, neat!

Reply:
It might be best to re-introduce the GenericEnvironment parameter to checkGenericArguments() since this comes up in a few places now.

Reply:
Perhaps. I can take care of that once I rebase 👍